packages/theme-language-server-common/src/completions/providers/HtmlAttributeCompletionProvider.ts

```diff
@@ -32,7 +41,49 @@ export class HtmlAttributeCompletionProvider implements Provider {
     const name = node.value;
     const partial = name.replace(CURSOR, '');
     const options = getOptions(partial, grandParentNodeName);
-    return options.sort(sortByName).map(toCompletionItem);
+
+    const attributeTagRange = this.attributeTagRange(node, document);
+    const hasExistingAttributeValue = this.hasExistingAttributeValue(attributeTagRange, document);
+    const hasLiquidTag = this.hasLiquidTag(attributeTagRange, document);
+
+    return options.sort(sortByName).map((tag) => {
+      return toCompletionItem(tag, attributeTagRange, hasExistingAttributeValue, hasLiquidTag);
+    });
+  }
+
+  hasExistingAttributeValue(attributeTagRange: Range, document: AugmentedSourceCode): boolean {
+    return /^\s*=/.test(
+      document.source.slice(document.textDocument.offsetAt(attributeTagRange.end)),
+    );
+  }
+
+  hasLiquidTag(attributeTagRange: Range, document: AugmentedSourceCode): boolean {
+    return /^(?:\{%|\{\{)/.test(
+      document.source.slice(document.textDocument.offsetAt(attributeTagRange.end)),
+    );
+  }
+
+  // Find the range of the attribute partial. If the attribute contains any liquid code, the range
+  // will end before the first character of the liquid block.
+  attributeTagRange(node: LiquidHtmlNode, document: AugmentedSourceCode): Range {
+    if (node.type === 'TextNode' && node.value === CURSOR) {
+      // If you try to auto-complete with no provided attribute tag,
+      // we will not try to override the subsequent character.
+      // E.g. <a href="" █>
+      return {
+        start: document.textDocument.positionAt(node.position.start),
+        end: document.textDocument.positionAt(node.position.start),
+      };
+    }
+
+    let attributeEndOffset =
+      document.source.slice(node.position.end).match(/[\s=]|{%|{{|>/)?.index ??
+      document.source.length;
```

Comment:
The reason i chose offset to be `document.source.length` instead of `0` is for this specific edge case:

```
<script d█efer
```

When we are trying to find the end of an attribute, we rely on space-like character, `=`, `{%`, or `{{`. The example above has none of those; the end of the document IS the attribute.

Im not sure there are any other ways i can do this 🤔

Reply:
OH i think i see what you mean...

It should probably be this

```
    const sourcePartialPastCursor = document.source.slice(node.position.end);

    let attributeEndOffset =
      sourcePartialPastCursor.match(/[\s=]|{%|{{|>/)?.index ??
      sourcePartialPastCursor.length;
```